New service: Logseq (#1052)
ralphbean/bugwarrior: bugwarrior/docs/services/logseq.rst

Logseq
======

You can import `tasks <https://docs.logseq.com/#/page/tasks>`_ from `Logseq <https://logseq.com/>`_ using the ``logseq`` service name.

Additional Requirements
-----------------------

> Logseq application must active with the HTTP API server running for bugwarrior to connect.

Reviewer: ```suggestion
Logseq application must interact with the HTTP API server running for bugwarrior to connect.
```

Is this what you meant? Then again, this sentence might be superfluous since the next one says the same thing more clearly.

Author: What I meant was "Logseq application must **be** active". Your right, next line says much the same thing so this can be dropped.

To use bugwarrior to pull tickets from Logseq you need to enable the Logseq HTTP APIs server.
In Logseq go to **Settings** > **Features** and toggle the **HTTP APIs server** option.

Next select the **API** option in the top menu to configure authorization token, e.g

.. image:: pictures/logseq_token.png

Example Service
---------------

Here's an example of a Logseq target:

.. config::

    [my_logseq_graph]
    service = logseq

>     logseq.host = localhost

Reviewer: ```suggestion
```

We only put required configuration fields here and this seems like a default most people would be using. Please add a section below to mention how one can change the `host` value.

Author: Moved to new section as suggested

    logseq.port = 12315
    logseq.token = mybugwarrioraccesstoken

The above example is the minimum required to import issues from Logseq.
You can also feel free to use any of the configuration options described in
:ref:`common_configuration_options` or described in `Service Features`_ below.

Service Features
----------------

Task filters
++++++++++++

You can configure the service to import tasks in different states.
By default the service will import all tasks in an active tasks states

    DOING, TODO, NOW, LATER, IN-PROGRESS, WAIT, WAITING

> You can override this filter by setting the ``logseq.task_state`` option to a 

Reviewer: ```suggestion
You can override this filter by setting the ``task_state`` option to a 
```

Best to leave out the service prefix in prose since those are not used in TOML. (In code blocks, the service prefixes are automatically removed when one clicks on the toml tab.)

comma separated list of required task states.

.. config::
    :fragment: logseq

>     logseq.task_state = DOING, NOW, IN-PROGRESS

Reviewer: We generally prefer adding a single `query` configuration option rather than building the query out of many different configuration options. Is there a reason that approach wouldn't work here?

For example:

```py
class LogseqConfig(config.ServiceConfig):
    ...
    query = ":where [?b :block/marker ?marker] [(contains? #{{\"DOING TODO NOW LATER IN-PROGRESS WAIT WAITING\"}} ?marker)]"
```

Author: Simplicity, and to avoid query issues. Arbitrary datascript queries can return result sets that do not match the expected list of tasks, plus issues like above where the `logseq.DB.datascriptQuery` function doesn't match to queries that work in the Logseq app itself.

Reviewer: > Simplicity...

That is arguable. If `task_state` were the only thing one might want to configure it is certainly simpler than a `query` option, but our experience with other services is that folks are going to want to configure every aspect of a query which leads to a much more complicated outcome than a single `query` parameter.

> Arbitrary datascript queries can return result sets that do not match the expected list of tasks,

Not all of our services map to things that are strictly "tasks" or "issues". For example, the Gmail service fetches emails. It seems to me that folks are likely to want to track artifacts in Logseq that are not "Logseq tasks".

> plus issues like above where the `logseq.DB.datascriptQuery` function doesn't match to queries that work in the Logseq app itself.

Is the `logseq.DB.datascriptQuery` API documented? Seems like it would be fine to say that the `query` option takes a `datascriptQuery` and link to the docs.

*BTW, I'm not hard set against your current approach, I'm just trying to draw out your rationale.*

Author: The API is *not* very well documented https://plugins-doc.logseq.com/logseq/DB/datascriptQuery
I got it working through trial and error.

I'd be happy to add a full query override as an advanced option, but maybe see if anyone actually asks for it first. I would suspect the simpler `task_state` selection will address 99%+ of users needs - if anyone other than me actually uses this integration :-)

Reviewer: > the simpler `task_state` selection will address 99%+ of users needs

Ok, that's basically what I was hoping to hear which leads into the next question i had: Is there anything special in Logseq about the default `task_state` values you've set or are they arbitrary strings used by convention? That is, why do you suspect users will want to change this value?

Author: The complete set of available task states in LogSeq is 
```
DOING, TODO, NOW, LATER, IN-PROGRESS, WAIT, WAITING, CANCELLED, CANCELED, DONE
```
I picked the set of what would be considered "open" tasks as the default.  My for my personal use I filter to just getting DOING, TODO, and WAITING

Reviewer: Can you explain your use case for wanting to exclude certain "open" states?

Author: LogSeq has two task workflow options TODO -> DOING -> DONE and LATER -> NOW -> DONE. I actually use a mix of both approaches where I tag things "I" need to do with TODO and DOING, and things I've delegated to others with NOW/LATER.  In taskwarrior I just want to pull the things I need to work on.

Reviewer: Thanks, that's the last piece of the puzzle. To summarize my understanding of this thread:

- The meaning of Logseq's states is ill-defined, so we need to provide a configuration mechanism for filtering out states that users may use for unactionable tasks.
- Mastery of the poorly documented datascriptQuery language is not necessary for all Logseq users, so merely exposing the entire query as a configuration option is not very user friendly.

Time will tell if you made the right judgment but I'm satisfied that you thought through this decision, so I'm going to mark this conversation as resolved.

Priority mapping
++++++++++++++++

> Logseq task priorities ``A``, ``B``, and ``C`` mapped to the taskwarrior priorities

Reviewer: ```suggestion
Logseq task priorities ``A``, ``B``, and ``C`` are mapped to the taskwarrior priorities
```

``H``, ``M``, and ``L`` respectively.

Task state and data/time mappings
+++++++++++++++++++++++++++++++++

``DOING``, ``TODO``, ``NOW``, ``IN-PROGRESS`` are mapped to the default ``pending`` state.
The Logseq task ``SCHEDULED:`` and ``DEADLINE:`` fields are mapped to the ``scheduled`` and 
``due`` date fields.

``LATER``, ``WAITING``, ``WAIT`` are mapped to the ``waiting`` state.

> The ``SHEDULED:`` date or ``DEADLINE`` date is used to set the ``wait`` date on the task.

Reviewer: ```suggestion
The ``SCHEDULED:`` date or ``DEADLINE`` date is used to set the ``wait`` date on the task.
```

If no scheduled or deadlines date is available then the wait date is set to ``someday`` 
(see ``Date and Time Synonyms <https://taskwarrior.org/docs/dates/#synonyms-hahahugoshortcode30s0hbhb/>``_).
Waiting tasks can be listed using ``task waiting``

``DONE`` is mapped to the ``completed`` state.

``CANCELED`` and ``CANCELLED`` are mapped to the ``deleted`` state.

Character replacement
+++++++++++++++++++++

> taskwarrior encodes the ``[`` and ``]`` characters commonly used in Logseq as ``&open;`` and ``&close;``. To

Reviewer: ```suggestion
Taskwarrior encodes the ``[`` and ``]`` characters commonly used in Logseq as ``&open;`` and ``&close;``. To
```

avoid display issues ``[[`` and ``]]`` are replaced by ``【`` and ``】`` for page links, and single
``[`` and ``]`` are replaced by ``〈`` and ``〉``. 

> You can override behaviour and use customer characters by setting the ``logseq.char_*`` options in your

Reviewer: ```suggestion
You can override behaviour and use customer characters by setting the ``char_*`` options in your
```

> ``bugwarriorrc`` config.

Reviewer: ```suggestion
You can override this behavior and use custom characters by setting the ``logseq.char_*`` options.
```

Since one can also use bugwarrior.toml it's best to avoid naming the file.

.. config::
    :fragment: logseq

    logseq.char_open_link = 〖
    logseq.char_close_link = 〗
    logseq.char_open_bracket = (

>     logseq.char_close_bracket = )

Reviewer: I'm not sure I understand the problem this feature is addressing. Why would regular square brackets cause display issues? Does the Logseq application render them differently?

Reviewer: I think [this comment](https://github.com/GothenburgBitFactory/bugwarrior/pull/1052#discussion_r1650043792) belongs here:

> This is a work around for what is possibly an underlying bug in taskwarrior. The `[` and `]` characters in the task description, and other fields, are replaced with `&open;` and `&close;`.

Hmm, at a cursory glance it seems like Taskwarrior is correctly encoding them but failing to decode in `src/Task.cpp`:

```cpp
////////////////////////////////////////////////////////////////////////////////
// Encode values prior to serialization.
//   [  -> &open;
//   ]  -> &close;
const std::string Task::encode (const std::string& value) const
{
  auto modified = str_replace (value,    "[", "&open;");
  return          str_replace (modified, "]", "&close;");
}

////////////////////////////////////////////////////////////////////////////////
// Decode values after parse.
//   [  <- &open;
//   ]  <- &close;
const std::string Task::decode (const std::string& value) const
{
  if (value.find ('&') == std::string::npos)
    return value;

  auto modified = str_replace (value,    "&open;",  "[");
  return          str_replace (modified, "&close;", "]");
}
```

@scross01 What version of taskwarrior are you on?

@djmitche Any thoughts on this potential Taskwarrior bug?

Author: I'm using taskwarrior 3.0.2 on MacOS, installed using Homebrew

Reviewer: I cannot reproduce the decoding bug on the command line by adding a task with brackets in the description on either tw-2 or tw-3. I'm not sure why bugwarrior would be any different since it is using the command line interface under the hood. @scross01 Can you verify manually adding and reading a task on the command line with brackets is or is not buggy on your system? Also, are you seeing the bug with the default report (e.g., running `task 42`) or are you using something fancier?

Reviewer 2: This has been a long-standing bug, which I thought I reported once upon a time but can't find in https://github.com/GothenburgBitFactory/taskwarrior/issues right now. I filed https://github.com/GothenburgBitFactory/taskwarrior/issues/3524 just now.

Author: I ran
```shell
task add "test [[Task]]"
```
and running `task list` returns the result with
```
22 1min                                         test [[Task]]                                           0
```

I removed the escape code from the integration, with a task in Logseq called 
```
TODO Test Logseq [[Task]]
```
then ran `bugwarrior pull` (note was previously synced with the escaping in place)
```
INFO:bugwarrior.db:Updating task a633286d-c62c-4ec5-88c0-669ff29a2e3a, (bw)Is#126 - TODO Test Logseq [[Task]]; logseqid: '126' -> 126; logseqtitle: 'TODO Test Logseq 〖Task〗' -> 'TODO Test Logseq [[Task]]'; description: '(bw)Is#126 - TODO Test Logseq 〖Task〗' -> '(bw)Is#126 - TODO Test Logseq [[Task]]'
``` 
and then `task list` includes
```
23 1min   MyLogseqGraph                                    (bw)Is#126 - TODO Test Logseq &open;&open;Task&close;&close;        1
```

Reviewer 2: Can you see what `task` invocation bugwarrior-pull is running in that case?

Author: Looks like the problem could be within the `taskw` library. I ran a simple test
```python
from taskw import TaskWarrior
w = TaskWarrior()
w.task_add("Task added using [[taskw]]")
```
```python
{'id': 24,
 'description': 'Task added using &open;&open;taskw&close;&close;',
 'entry': '20240628T102423Z',
 'modified': '20240628T102423Z',
 'status': 'pending',
 'uuid': '9f1544f9-e66e-41ca-b65e-79c328bac19c',
 'urgency': 0}
```

Author: specifically this appears to be the cause https://github.com/ralphbean/taskw/blob/3256c92bed4bb3d53269c843e8201e9ac65780cc/taskw/utils.py#L27

`taskw` is escaping the `[`, `]` and `"` characters before invoking `task` command line.
```python
[b'task', b'rc.verbose=new-uuid', b'rc.json.array=TRUE', b'rc.confirmation=no', b'rc.dependency.confirmation=no', b'rc.recurrence.confirmation=no', b'9f1544f9-e66e-41ca-b65e-79c328bac19c', b'modify', b'description:"Task added using &open;&open;taskw&close;&close; + (does) {this} &dquot;quote&dquot; work?"', b'entry:"20240628T102423Z"', b'modified:"20240628T110925Z"', b'status:"pending"']
```

Reviewer 2: Thanks! I've updated the taskwarrior bug.

Reviewer: Created https://github.com/ralphbean/taskw/issues/172 to track the actual underlying bug in taskw. I'm thinking if we documented at the beginning of this section that this is a workaround for https://github.com/ralphbean/taskw/issues/172 to give a little more context to these options that would resolve this thread.

Author: Updated with reference to the `taskw` issue in the code and the service doc.

Reviewer: Thanks. Maybe I'm reading too much into it but I'm curious if there was intent behind the phrase "in part" where you say "This capability is *in part* to workaround...". I presume you'd be fine with deprecating and eventually removing this feature once the taskw bug is fixed?

Author: Specifically for the double braces `[[` and `]]` I would consider leaving the capability in place, to allow alternate styling of the page references.

Reviewer 2: It would be great to update the comment to say that `taskw` does this encoding, not Taskwarrior.

Author: Already updated in commit https://github.com/GothenburgBitFactory/bugwarrior/pull/1052/commits/7a94d57f663b89b8445360e234772c0c1f75deb3

Reviewer 2: Oh, sorry, the GitHub UI confused me

Reviewer: > Specifically for the double braces `[[` and `]]` I would consider leaving the capability in place, to allow alternate styling of the page references.

Does Logseq itself have a feature to change this styling? To me this would seem like a vanity feature which doesn't provide enough value to justify its maintenance and documentation if it were not for the taskw bug it serves to work around.

*Note: Resolving this difference of opinion is not blocking this PR at this point and can be postponed until the upstream bug is fixed.*

Logseq URI links
++++++++++++++++

A ``logseq://`` URI is generated for each task to enable easy navigation directly to the specific task in
the Logseq application. 

By default bugwarrior incorporates the links into task description. To disable this behaviour either 
modify the ``inline_links`` option to affect all services, or to modify for the logseg sevice only you can 

> set the ``logseq.inline_links`` option to False in your ``bugwarriorrc``.

Reviewer: ```suggestion
modify the ``inline_links`` option in the main section to affect all services, or to modify for the logseg sevice only you can 
set it in your Logseq section.
```

Logseq URI in MacOS add the following to your ``~/..zshrc``

.. code-block:: bash

    # open a specific taskwarrior task in Logseq
    function taskopen() {
        open $(task $1 | grep "Logseq URI" | sed -r 's/^Logseq URI//')
    }

From the command line you can open a specific task using taskwarior task id, e.g. ``taskopen 1234``.

Troubleshooting---------------

Logseq graph re-index
+++++++++++++++++++++

If you re-index your Logseq graph all task ids and uuids are changed. The next time
you run bugwarrior all existing taskwarrior tasks will be closed and new ones will 
be created.

Logseq API connection issues
++++++++++++++++++++++++++++

If you get the following error when running bugwarrior:

    CRITICAL:bugwarrior.services.logseq:Unable to connect to Logseq HTTP APIs server. HTTPConnectionPool(host='localhost', port=12315): Max retries exceeded with url: /api (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x105764f20>: Failed to establish a new connection: [Errno 61] Connection refused'))

- Check that the LogSeq application is running
- Check that the HTTP APIs server is started

> - Check that authorization token is set the APIs server settings and matches the 

Reviewer: ```suggestion
- Check that authorization token is set in the API server settings and matches the 
```

>   ``logseq.token`` in your ``bugwarriorrc`` 

Reviewer: ```suggestion
  ``token``.
```

+++++++++++++++++++

The authorization token is used to authenticate with Logseq. This value is required and must match
the one of the authorization tokens set in Logseq HTTP APIs server settings.

.. config::
    :fragment: logseq

    logseq.token = mybugwarrioraccesstoken

Task filters
++++++++++++

You can configure the service to import tasks in different states.
By default the service will import all tasks in an active tasks states

    DOING, TODO, NOW, LATER, IN-PROGRESS, WAIT, WAITING

You can override this filter by setting the ``task_state`` option to a 
comma separated list of required task states.

.. config::
    :fragment: logseq

    logseq.task_state = DOING, NOW, IN-PROGRESS

Task state and data/time mappings
+++++++++++++++++++++++++++++++++

``DOING``, ``TODO``, ``NOW``, ``LATER``, and ``IN-PROGRESS`` are mapped to the default ``pending`` state.
The Logseq task ``SCHEDULED:`` and ``DEADLINE:`` fields are mapped to the ``scheduled`` and 
``due`` date fields.

``WAITING`` and ``WAIT`` are dynamically mapped to either ``pending`` or ``waiting`` states based on 
the ``wait`` date. The ``SCHEDULED:`` date or ``DEADLINE`` date is used to set the ``wait`` date on the

> task. If no scheduled or deadlines date is available then the wait date is set to ``someday`` 

Reviewer: ```suggestion
task. If no scheduled or deadline date is available then the wait date is set to ``someday`` 
```

Author: fixed
